File: src/main/java/com/lalkalol/mujahid/audio/LavalinkManager.java
package com.lalkalol.mujahid.audio;

import com.lalkalol.mujahid.config.Config;
import dev.arbjerg.lavalink.client.LavalinkClient;
import dev.arbjerg.lavalink.client.NodeOptions;
import dev.arbjerg.lavalink.client.event.ReadyEvent;
import dev.arbjerg.lavalink.client.event.StatsEvent;
import dev.arbjerg.lavalink.client.event.TrackEndEvent;
import dev.arbjerg.lavalink.client.event.TrackStartEvent;
import dev.arbjerg.lavalink.client.event.WebSocketClosedEvent;
import dev.arbjerg.lavalink.client.loadbalancing.builtin.VoiceRegionPenaltyProvider;
import net.dv8tion.jda.api.JDA;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.ConcurrentHashMap;

import dev.arbjerg.lavalink.client.Helpers;

public class LavalinkManager {
    private static final int SESSION_INVALID = 4006;
    private static final Logger log = LoggerFactory.getLogger(LavalinkManager.class);

    private final Config config;
    private final LavalinkClient client;
    private final ConcurrentHashMap<Long, GuildMusicManager> musicManagers = new ConcurrentHashMap<>();

    private volatile JDA jda;

    public LavalinkManager(Config config) {
        this.config = config;
        this.client = new LavalinkClient(Helpers.getUserIdFromToken(config.discordToken()));
    }

    public LavalinkClient getClient() {
        return client;
    }

    public JDA getJda() {
        return jda;
    }

    public void setJda(JDA jda) {
        this.jda = jda;
    }

    public void start() {
        client.getLoadBalancer().addPenaltyProvider(new VoiceRegionPenaltyProvider());
        registerEventListeners();
        addNode();
    }

    public GuildMusicManager getOrCreate(long guildId) {
        return musicManagers.computeIfAbsent(guildId, id -> new GuildMusicManager(id, this));
    }

    public GuildMusicManager getExisting(long guildId) {
        return musicManagers.get(guildId);
    }

    public void destroy(long guildId) {
        musicManagers.remove(guildId);
        log.debug("Destroyed music manager for guild {}", guildId);
    }

    public void shutdown() {
        log.info("Shutting down Lavalink client");
        client.close();
    }

    private void addNode() {
        client.addNode(
                new NodeOptions.Builder()
                        .setName("main")
                        .setServerUri(config.lavalinkUri())
                        .setPassword(config.lavalinkPassword())
                        .build()
        );
    }

    private void registerEventListeners() {
        client.on(ReadyEvent.class).subscribe(event ->
                log.info("Lavalink node '{}' ready (session {})", event.getNode().getName(), event.getSessionId())
        );

        client.on(StatsEvent.class).subscribe(event ->
                log.debug(
                        "Node '{}' players {}/{}",
                        event.getNode().getName(),
                        event.getPlayingPlayers(),
                        event.getPlayers()
                )
        );

        client.on(TrackStartEvent.class).subscribe(event -> {
            GuildMusicManager manager = getExisting(event.getGuildId());
            if (manager != null) {
                manager.getScheduler().onTrackStart(event.getTrack());
            }
        });

        client.on(TrackEndEvent.class).subscribe(event -> {
            GuildMusicManager manager = getExisting(event.getGuildId());
            if (manager != null) {
                manager.getScheduler().onTrackEnd(event.getTrack(), event.getEndReason());
            }
        });

        client.on(WebSocketClosedEvent.class).subscribe(event -> {
            if (event.getCode() == SESSION_INVALID) {
                reconnectVoice(event.getGuildId());
            }
        });
    }

    private void reconnectVoice(long guildId) {
        if (jda == null) {
            return;
        }
        var guild = jda.getGuildById(guildId);
        if (guild == null) {
            return;
        }
        var channel = guild.getSelfMember().getVoiceState().getChannel();
        if (channel == null) {
            return;
        }
        log.warn("Voice session invalid for guild {}, reconnecting", guildId);
        guild.getJDA().getDirectAudioController().reconnect(channel);
    }
}

File: src/main/kotlin/com/lalkalol/mujahid/audio/LavalinkManager.kt
package com.lalkalol.mujahid.audio

import com.lalkalol.mujahid.config.Config
import dev.arbjerg.lavalink.client.LavalinkClient
import dev.arbjerg.lavalink.client.getUserIdFromToken
import dev.arbjerg.lavalink.client.NodeOptions
import dev.arbjerg.lavalink.client.event.ReadyEvent
import dev.arbjerg.lavalink.client.event.StatsEvent
import dev.arbjerg.lavalink.client.event.TrackEndEvent
import dev.arbjerg.lavalink.client.event.TrackStartEvent
import dev.arbjerg.lavalink.client.event.WebSocketClosedEvent
import dev.arbjerg.lavalink.client.loadbalancing.builtin.VoiceRegionPenaltyProvider
import net.dv8tion.jda.api.JDA
import org.slf4j.LoggerFactory
import java.util.concurrent.ConcurrentHashMap

class LavalinkManager(private val config: Config) {
    private val log = LoggerFactory.getLogger(LavalinkManager::class.java)
    val client: LavalinkClient = LavalinkClient(getUserIdFromToken(config.discordToken))
    private val musicManagers = ConcurrentHashMap<Long, GuildMusicManager>()

    @Volatile
    var jda: JDA? = null
        private set

    fun setJda(jda: JDA) {
        this.jda = jda
    }

    fun start() {
        client.loadBalancer.addPenaltyProvider(VoiceRegionPenaltyProvider())
        registerEventListeners()
        addNode()
    }

    fun getOrCreate(guildId: Long): GuildMusicManager =
        musicManagers.computeIfAbsent(guildId) { GuildMusicManager(it, this) }

    fun getExisting(guildId: Long): GuildMusicManager? = musicManagers[guildId]

    fun destroy(guildId: Long) {
        musicManagers.remove(guildId)
        log.debug("Destroyed music manager for guild {}", guildId)
    }

    fun getAllManagers(): List<Map.Entry<Long, GuildMusicManager>> =
        musicManagers.entries.toList()

    fun activePlayerCount(): Int = musicManagers.size

    fun shutdown() {
        log.info("Shutting down Lavalink client")
        client.close()
    }

    private fun addNode() {
        client.addNode(
            NodeOptions.Builder()
                .setName("main")
                .setServerUri(config.lavalinkUri())
                .setPassword(config.lavalinkPassword)
                .build(),
        )
    }

    private fun registerEventListeners() {
        client.on(ReadyEvent::class.java).subscribe { event ->
            log.info(
                "Lavalink node '{}' ready (session {})",
                event.node.name,
                event.sessionId,
            )
        }

        client.on(StatsEvent::class.java).subscribe { event ->
            log.debug(
                "Node '{}' players {}/{}",
                event.node.name,
                event.playingPlayers,
                event.players,
            )
        }

        client.on(TrackStartEvent::class.java).subscribe { event ->
            getExisting(event.guildId)?.scheduler?.onTrackStart(event.track)
        }

        client.on(TrackEndEvent::class.java).subscribe { event ->
            getExisting(event.guildId)?.scheduler?.onTrackEnd(event.track, event.endReason)
        }

        client.on(WebSocketClosedEvent::class.java).subscribe { event ->
            if (event.code == SESSION_INVALID) {
                reconnectVoice(event.guildId)
            }
        }
    }

    private fun reconnectVoice(guildId: Long) {
        val jdaInstance = jda ?: return
        val guild = jdaInstance.getGuildById(guildId) ?: return
        val channel = guild.selfMember.voiceState?.channel ?: return
        log.warn("Voice session invalid for guild {}, reconnecting", guildId)
        guild.jda.directAudioController.reconnect(channel)
    }

    companion object {
        private const val SESSION_INVALID = 4006
    }
}
